bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/common/SafeCallManagerImpl.java

```diff
@@ -0,0 +1,168 @@
+            if (enforceSingleThreadPerIdentifier && otherWrapper != null) {
+                // another call to the same identifier is (still) running,
+                // therefore queue it instead for async execution later on.
+                // Inform the caller about the timeout by means of the exception.
+                enqueue(invocation);
+                throw new DuplicateExecutionException(otherWrapper);
+            }
+            activeIdentifiers.put(invocation.getIdentifier(), wrapper);
+        }
+        if (invocation.getInvocationHandler() instanceof InvocationHandlerAsync) {
+            watch(invocation);
+        }
+    }
+
+    @Override
+    public void recordCallEnd(Invocation invocation, TrackingCallable wrapper) {
+        synchronized (activeIdentifiers) {
+            activeIdentifiers.remove(invocation.getIdentifier());
+        }
+        synchronized (activeAsyncInvocations) {
+            activeAsyncInvocations.remove(invocation.getIdentifier());
+        }
+        logger.trace("Finished {}", invocation);
+        trigger(invocation.getIdentifier());
+    }
+
+    @Override
+    public void enqueue(Invocation call) {
+        synchronized (queues) {
+            Queue<Invocation> queue = queues.get(call.getIdentifier());
+            if (queue == null) {
+                queue = new LinkedList<>();
+                queues.put(call.getIdentifier(), queue);
+            }
+            queue.add(call);
+        }
+        trigger(call.getIdentifier());
+    }
+
+    private void trigger(Object identifier) {
+        logger.trace("Triggering submissions for '{}'", identifier);
+        synchronized (activeIdentifiers) {
+            if (enforceSingleThreadPerIdentifier && activeIdentifiers.containsKey(identifier)) {
+                logger.trace("Identifier '{}' is already running", identifier);
+                return;
+            }
+        }
+        synchronized (activeAsyncInvocations) {
+            if (activeAsyncInvocations.containsKey(identifier)) {
+                logger.trace("Identifier '{}' is already scheduled for asynchronous execution", identifier);
+                return;
+            }
+            Invocation next = dequeue(identifier);
+            if (next != null) {
+                logger.trace("Scheduling {} for asynchronous execution", next);
+                activeAsyncInvocations.put(identifier, next);
+                TrackingCallable wrapper = new TrackingCallable(next);
+                getScheduler().submit(wrapper);
+                logger.trace("Submitted {} for asynchronous execution", next);
+            }
+        }
+    }
+
+    private void handlePotentialTimeout(Invocation invocation) {
+        Object identifier = invocation.getIdentifier();
+        Invocation currentInvocation = activeAsyncInvocations.get(identifier);
+        if (currentInvocation != null && currentInvocation == invocation) {
```

Comment:
`invocation` is not null as it has been already dereferenced above.
So, if you test `currentInvocation == invocation`, isn't the test `currentInvocation != null` redundant?